File: 2018/python/solution1.py
"""
Advent of code day 1
"""

def find_endstate(fname):
    total = 0
    with open(fname, 'r') as fin:
        for line in fin:
            total += int(line)
    return total


def find_repeat(fname):
    with open(fname, 'r') as fin:
        data = [int(line) for line in fin]
    total = 0
    seen = set([total])
    while True:
        for d in data:
            total += d
            if total in seen:
                return total
            seen.add(total)

print("First repeat = {}".format(find_repeat("../data/input-1.txt")))

File: 2018/nim/solution1.nim
"""
Advent of code day 1
"""
import strutils, sequtils, intsets

proc parse_file(fname: string): seq[int] =
  for line in open(fname).lines:
    result.add(line.parseInt)

func sum[T](x: seq[T]): T =
  for elem in x:
    result += elem

proc find_endstate(fname: string): int =
  return parse_file(fname).sum

proc find_repeat(fname: string): int =
  let input = parse_file(fname)
  var seen = initIntSet()
  seen.incl(result)
  while true:
    for line in input:
      result += line
      if seen.containsOrIncl(result):
        return result

echo find_endstate("../data/input-1.txt")
echo find_repeat("../data/input-1.txt")
